alternator/streams.cc

```diff
@@ -491,34 +508,107 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
 
         std::optional<shard_id> last;
 
+        if (shard_start) {
+            i = std::find_if(i, e, [&](auto& p) {
+                return p.first == shard_start->start_time;
+            });
+        }
+
+        // filter out cdc generations older than the table or now() - 24h
+        auto low_ts = std::min(as_timepoint(schema->id()), db_clock::now() - 24h);
+        while (i != e) {
+            auto j = i;
+            if (++j != e && j->first < low_ts) {
+                i = j;
+                continue;
+            }
+            // next is younger.
+            break;
+        }
+
         for (; limit > 0 && i != e; prev = i, ++i) {
             auto& [ts, sv] = *i;
 
             last = std::nullopt;
 
-            for (auto& id : sv.streams()) {
-                if (shard_start && shard_start->id != id) {
-                    continue;
+            // TODO: get rid of this once cdc generations
+            // management starts correctly updating
+            // cdc expiration columns in topology/desc
+            auto get_expiration = [&](auto iter) {
+                auto exp = iter->second.expired();
+                if (!exp && ++iter != e) {
+                    exp = iter->first;
                 }
-                if (shard_start && shard_start->id == id) {
-                    shard_start = std::nullopt;
-                    continue;
+                return exp.value_or(db_clock::time_point{});
+            };
+
+            auto lo = sv.streams().begin();
+            auto end = sv.streams().end();
+
+            // operating on a copy. ok to sort. this order is also preserved 
+            // for the "next" check below.
+            std::sort(lo, end, [](auto& id1, auto& id2) {
+                auto i1 = id1.index();
+                auto i2 = id2.index();
+                return i1 < i2 || (i1 == i2 && id1.token() < id2.token());
+            });
+
+            assert(std::is_sorted(std::find_if(lo, end, [](auto& id) { return id.index() == 1; }), end, [](auto& id1, auto& id2) {
+                return id1.token() < id2.token();
+            }));
+
+            if (shard_start) {
+                lo = std::find_if(lo, end, [t = shard_start->hi_token](const cdc::stream_id& id) {
+                    return id.token() == t;
+                });
+                shard_start = std::nullopt;
+            }
+
+            for (auto hi = lo; lo != end; lo = hi) {
+                auto& id = *lo++;
+
+                hi = std::find_if(lo, end, [i = id.index()](const cdc::stream_id& id) {
+                    return id.index() != i;
+                });
+
+                auto t = id.token();
+
+                auto et = hi != end ? hi->token() : dht::maximum_token();
+                if (et < t) {
+                    assert(id.index() == 0);
+                    // wraparound id. index 0. just treat is as (min, end] instead.
```

Comment:
Oh, I think I got my answer to the wraparound question above.

Reply:
Yes. We generate two ranges that are not fully in line with what the stream ID says, (min, lo] (hi, max]. To be able to have sorted token ranges.